external_api/api_solargis.py

```diff
@@ -0,0 +1,271 @@
+
+    def get_arbitrary_payload(self, xml_request_content):
+        headers = {'Content-Type' : 'application/xml'}
+
+        try:
+            response = requests.post(
+                self.api_request_url,
+                params=self.api_key_params,
+                headers=headers,
+                data=xml_request_content.encode('utf8'),
+                timeout=30
+            )
+        except requests.Timeout:
+            logger.error("Request timeout")
+            return 408, None
+
+        if response.status_code != 200:
+            logger.error(response.text)
+            return response.status_code, None
+
+        text_response = response.text
+        return response.status_code, text_response
+
+    def text_response_to_readings(self, text_response, request_time=None):
+
+        request_time = request_time or datetime.datetime.now(datetime.timezone.utc)
+
+        response_dict = self.schema.to_dict(text_response)
+        readings_dirty = response_dict['site'][0]['row']
+        readings = [
+            (rfc3336todt(v['@dateTime']), *v['@values'], self.api_name, request_time)
+            for v in readings_dirty
+        ]
+        return readings
+
+    def get_current_solargis_irradiance_readings_site(self, site: Site, from_date, to_date, processing_keys):
+
+        from_date_str = datetime.datetime.strftime(from_date, '%Y-%m-%d')
+        to_date_str = datetime.datetime.strftime(to_date, '%Y-%m-%d')
+
+        xml_request_content = f'''
+            <ws:dataDeliveryRequest dateFrom="{from_date_str}" dateTo="{to_date_str}"
+            xmlns="http://geomodel.eu/schema/data/request"
+            xmlns:ws="http://geomodel.eu/schema/ws/data"
+            xmlns:geo="http://geomodel.eu/schema/common/geo"
+            xmlns:pv="http://geomodel.eu/schema/common/pv"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+            <site id="{site.name}" lat="{site.latitude}" lng="{site.longitude}">
+             <pv:system installedPower="{site.peak_power_w}">
+                <pv:module type="ASI"/>
+                <pv:inverter/>
+                <pv:losses/>
+             </pv:system>
+            </site>
+            <processing key="{processing_keys}" summarization="HOURLY" terrainShading="true">
+            </processing>
+            </ws:dataDeliveryRequest>
+        '''
+
+        logger.debug(xml_request_content)
+
+        status, text_response = self.get_arbitrary_payload(xml_request_content)
+
+        readings = self.text_response_to_readings(text_response) if status == 200 else None
+
+        return status, readings
+
+    def get_current_solargis_irradiance_readings(self, from_date=None, to_date=None, processing_keys=None):
+
+        processing_keys = processing_keys or 'GHI GTI TMOD PVOUT'
+        from_date = from_date or datetime.date.today() - datetime.timedelta(days=1)
+        to_date = to_date or datetime.date.today() - datetime.timedelta(days=1)
```

Comment:
⚠️

from_date = to_date si no li entres per paràmetre!

Crec que no hi ha test d'això, en tots se li passa un from i un to.

Reply:
Estem constringint a date i from_date = to_date per un dia (que serà la tasca diaria per defecte) torna 24h